staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go

```diff
@@ -134,10 +136,29 @@ func (r *proxyHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 	location.Scheme = "https"
 	rloc, err := r.serviceResolver.ResolveEndpoint(handlingInfo.serviceNamespace, handlingInfo.serviceName, handlingInfo.servicePort)
 	if err != nil {
-		klog.Errorf("error resolving %s/%s: %v", handlingInfo.serviceNamespace, handlingInfo.serviceName, err)
+		utilruntime.HandleError(fmt.Errorf("error resolving %s/%s: %w", handlingInfo.serviceNamespace, handlingInfo.serviceName, err))
 		proxyError(w, req, "service unavailable", http.StatusServiceUnavailable)
 		return
 	}
+
+	// If the ServiceResolver also implements ServiceHostnameResolver, check to see if an external hostname should be used for TLS config.
+	if hostnameResolver, ok := r.serviceResolver.(ServiceHostnameResolver); ok {
+		hostname, isExternal, err := hostnameResolver.ResolveHostname(handlingInfo.serviceNamespace, handlingInfo.serviceName)
+		if err != nil {
+			utilruntime.HandleError(fmt.Errorf("error resolving hostname %s/%s: %w", handlingInfo.serviceNamespace, handlingInfo.serviceName, err))
+			proxyError(w, req, "hostname unavailable", http.StatusServiceUnavailable)
+			return
+		}
+		if isExternal {
+			handlingInfo.transportConfig.TLS.ServerName = hostname
```

Comment:
do we have e2e tests with external services? Would we notice if this code block is wrong?

Reply:
> do we have e2e tests with external services?

i don't believe so, but i could add them. this is my first PR, so i am probably missing a lot of context about e2e. do the e2e tests run against a full kubernetes cluster? if so, i might need some guidance with how to launch an apiserver outside of the cluster

> Would we notice if this code block is wrong?

i have added tests in `handler_proxy_test.go` that seem to cover this scenario. could you clarify what you mean here?